### PR TITLE
Np 25966 biobank module

### PR DESCRIPTION
**Pull request by tkardu**

### CODE REVIEW

**Comment by ketilaa** [comment]
I see that you are still working on the module branch. You should have merged it to main when it was approved, and created a new one for the model part. But I know getting coverage before you have implemented everything might be hard. You temporarily lower the coverage limit for modules that are under development to be able to merge to main still. Up to you.

File: cristin-biobank/src/main/src/java/no/unit/nva/model/nva/Biobanks.java
Comment (ketilaa) [inline cristin-biobank/src/main/src/java/no/unit/nva/model/nva/Biobanks.java:34]: When looking at the json representation, biobanks looks better than biobankList. Put if you use biobankList, be consistent and give the constructor parameter the same name.

**Review by ketilaa — CHANGES_REQUESTED** [review]
Looking good. Still fields missing that need to be added :)

File: cristin-commons/src/main/java/no/unit/nva/utils/DateInfo.java
Comment (torarnet) [inline cristin-commons/src/main/java/no/unit/nva/utils/DateInfo.java:1]: This is a model class, not a util class. I think a better place would be to put DateInfo in model package, where Organization.java is. The module seems right. Things we need for more than one module should be in commons module, as you have done here

File: cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java:71]: Is there i reason not to use typed list? But its not a big issue

File: cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java:48]: Are these null safe? Do we now that identifier from model is always non null?

File: cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java:59]: null safe?

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinAssocProjectForBiobank.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinAssocProjectForBiobank.java:29]: We might want to consider for later annotating with JsonCreator and adding the JsonProperty here as well, but we dont need to do that yet until the need arrives.

Also for later, consider creating the toString, equals and hashcode if we need them

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java:18]: This is also a good way to define name for a constant. The important thing when giving names is we understand easily which constant map to each field.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java:42]: We already know that the field is a part of the class CristinBiobank, so I do not think it is necessary to prefix all field names in class CristinBiobank with cristinBiobank. Maybe i need a second opinion

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinCoordinatinInstitution.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinCoordinatinInstitution.java:7]: What about this:
https://github.com/BIBSYSDEV/nva-cristin-service/blob/main/cristin-commons/src/main/java/no/unit/nva/cristin/model/CristinOrganization.java
?

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinExternalSourcesBiobank.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinExternalSourcesBiobank.java:5]: https://github.com/BIBSYSDEV/nva-cristin-service/blob/main/cristin-project/src/main/java/no/unit/nva/cristin/projects/model/cristin/CristinExternalSource.java

Comment (torarnet) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinExternalSourcesBiobank.java:5]: You can either move that one to commons, or, if you feel there is better to have a duplicate in local module, maybe we should keep it as is

**Review by torarnet — COMMENTED** [review]
Some little things here and there. We can discuss tomorrow. I might need a second opinion if you can request review from more than me

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinAssocProjectForBiobank.java
Comment (torarnet) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinAssocProjectForBiobank.java:29]: Is also relevant other places, but not very important yet

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinCoordinatinInstitution.java
Comment (tkardu) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinCoordinatinInstitution.java:7]: Originally I was planning somehow to use it from commons,
But then Ketil said that in NVA model biobank we just need URI of the organization and unit + it is not all information about organization which is filled. 
Maybe it is more reasonable to use organization still, true. 
Need second opinion as you said.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java
Comment (tkardu) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java:42]: I did mostly for myself and future code exploration. We do know it, but when someone tries to navigate the code it makes it way easier to have it prefixed to avoid confusion?

**Review by brinxmat — CHANGES_REQUESTED** [review]
See other reviewer's comments

File: cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java
Comment (ketilaa) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java:71]: There is no good reason to use untyped List is most cases. Should be changed.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java
Comment (ketilaa) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java:42]: Well, class name is CristinBiobank. It is much easier to have a field named startDate(). All the extra chatter prefixing all fields with cristinBiobank is just noise in my opinion.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinCoordinatinInstitution.java
Comment (ketilaa) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinCoordinatinInstitution.java:7]: You can use CristinOrganization and populate only id (uri) too.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinExternalSourcesBiobank.java
Comment (ketilaa) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinExternalSourcesBiobank.java:5]: Why keep duplicates? If it is already in commons, use it.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java
Comment (tkardu) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java:71]: I have just fixed that.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinExternalSourcesBiobank.java
Comment (tkardu) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinExternalSourcesBiobank.java:5]: It is not in commons. I will move it there.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java:101]: You should annotate every parameter, not just the first 4

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java:113]: use parameter names matching field name, example:
`this.biobankId = biobankId`

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinAssocProjectForBiobank.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinAssocProjectForBiobank.java:14]: This points to string value "url" and is hard to understand. Use better names describing the constants.

```
public static final String CRISTIN_PROJECT_ID = "cristin_project_id";
public static final String CRISTIN_PROJECT_URL = "url";
public static final String CRISTIN_PROJECT_TITLE = "title";
```

Or something similar. See earlier comment

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/ExternalSourcesBiobank.java
Comment (torarnet) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/ExternalSourcesBiobank.java:10]: Ask other reviewer if we should use these names for the data model

**Review by torarnet — COMMENTED** [review]
See comments

File: cristin-biobank/src/main/java/no/unit/nva/biobank/common/DomainUriUtils.java
Comment (ketilaa) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/common/DomainUriUtils.java:44]: Url encoding the cristinBiobankId is not necessary as it can only contain digits.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java
Comment (ketilaa) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/common/MappingUtils.java:42]: You have some typos here I think. Is it not supposed to be `coordinatingInstitution`? Same applies to variable and getter in `CristinBiobank`.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/common/DomainUriUtils.java
Comment (ketilaa) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/common/DomainUriUtils.java:28]: Just wondering. Why do you need both `getBiobankUri()` and `getBiobankParamUri()`? Will the path to a biobank be different than `BIOBANK_PATH_ELEMENT` in any cases?

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java
Comment (ketilaa) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinBiobank.java:60]: `coordinatingInstitution`

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinCoordinatinInstitution.java
Comment (ketilaa) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/cristin/CristinCoordinatinInstitution.java:7]: `CristinCoordinatingInstitution`

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/Biobank.java
Comment (ketilaa) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/Biobank.java:40]: I think you should use transient only if there is no getter made available.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/Biobank.java
Comment (ketilaa) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/Biobank.java:119]: Be consistent in naming. `type` vs `biobankType`

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/ExternalSourcesBiobank.java
Comment (ketilaa) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/ExternalSourcesBiobank.java:10]: I would have called this class `ExternalSource`. And the fields `source` and `reference`. Is it a similar model elsewhere in the Cristin API?

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/TimeStampFromSource.java
Comment (ketilaa) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/TimeStampFromSource.java:8]: You call the class `TimeStampFromSource`, so fields can just be `timeStamp` and `source`. I would not have use camel-case for timeStamp, just timestamp.

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/BiobankMaterial.java
Comment (ketilaa) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/BiobankMaterial.java:8]: @torarnet I think description would be enough here, or?

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/BiobankMaterial.java
Comment (ketilaa) [inline cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/BiobankMaterial.java:21]: While this is ok, I would rather have created a static method for creating an instance of BiobankMaterial from a different object:

`
public static Biobank fromCristinBiobank(CristinBiobank cristinBiobank) {...}
`

Or extract this to some mapping utils if there is no reason why the classes should know about each other.

**Review by ketilaa — CHANGES_REQUESTED** [review]

File: cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/BiobankMaterial.java
Comment (torarnet) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/BiobankMaterial.java:8]: We should ask @brinxmat about domain model.

Also, here is a class for ExternalSource, found in project module:
https://github.com/BIBSYSDEV/nva-cristin-service/blob/main/cristin-project/src/main/java/no/unit/nva/cristin/projects/model/nva/ExternalSource.java

Comment (ketilaa) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/BiobankMaterial.java:8]: Looks like you can reuse ExternalSource from project module. Move it to commons.

Comment (ketilaa) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/BiobankMaterial.java:8]: Yes, it was you that commented on ExternalSource in BiobankMaterial comment :)

Comment (torarnet) [reply cristin-biobank/src/main/java/no/unit/nva/biobank/model/nva/BiobankMaterial.java:8]: I was not able to write a comment in the other comment, had to write a comment here instead. dont know why

**Pull request closed without merging**
